src/domain/chainContext.ts

```diff
@@ -183,15 +186,23 @@ export class ChainContext {
    * 1. Check if there are any `ConditionalOrderCreated` events, and index these.
    * 2. Check if any orders want to create discrete orders.
    */
-  public async runBlockWatcher() {
-    const { provider, chainId } = this;
+  private async runBlockWatcher() {
+    const { provider, registry, chainId } = this;
     const _ = (s: string) =>
       console.log(`[runBlockWatcher:chainId:${chainId}] ${s}`);
     // Watch for new blocks
     _("Subscribe to new blocks");
+    let lastBlockReceived = 0;
+    let timeLastBlockProcessed = new Date().getTime();
     provider.on("block", async (blockNumber: number) => {
       try {
         _(`New block ${blockNumber}`);
+        timeLastBlockProcessed = new Date().getTime();
+        if (blockNumber <= lastBlockReceived) {
+          // This may be a re-org, so process the block again
+          _(`Re-org detected, re-processing block ${blockNumber}`);
```

Comment:
Still using `_` as a fn name?

![](https://external-content.duckduckgo.com/iu/?u=http%3A%2F%2F31.media.tumblr.com%2F7b05db84d9f1bf2ba7bad72732e1421e%2Ftumblr_ml01gvUPCG1r18fjgo1_500.gif&f=1&nofb=1&ipt=bb42f2931b51a70ee40ced343801fc32af99a3934fd5deda6829b31704359169&ipo=images)

Reply:
Heh, yeah we have just settled on a debugging package and methodology, so all this is going to be given a good clean up in what will be a much anticipated PR :wink: